cvmfs/cvmfs_suid_helper.cc

```diff
@@ -104,64 +105,127 @@ static void KillCvmfs(const string &fqrn) {
   ExecAsRoot("/bin/kill", "-9", pid.c_str(), NULL);
 }
 
-static bool ClearWorkingDir() {
-  int retval;
-  DIR *dirp = opendir(".");
-  if (dirp == NULL)
-    return false;
-  platform_dirent64 *dirent;
-  while ((dirent = platform_readdir(dirp)) != NULL) {
-    if ((strcmp(dirent->d_name, ".") == 0) ||
-        (strcmp(dirent->d_name, "..") == 0))
-    {
-      continue;
+class ScopedWorkingDirectory {
+ public:
+  ScopedWorkingDirectory(const string &path)
+    : previous_path_(GetCurrentWorkingDirectory())
+    , directory_handle_(NULL)
+  {
+    ChangeDirectory(path);
+    directory_handle_ = opendir(".");
+  }
+
+  ~ScopedWorkingDirectory() {
+    if (directory_handle_ != NULL) {
+      closedir(directory_handle_);
+    }
+    ChangeDirectory(previous_path_);
+  }
+
+  operator bool() const { return directory_handle_ != NULL; }
+
+  struct DirectoryEntry {
+    bool   is_directory;
+    string name;
+  };
+
+  bool NextDirectoryEntry(DirectoryEntry *entry) {
+    platform_dirent64 *dirent;
+    while ((dirent = platform_readdir(directory_handle_)) != NULL &&
+           IsMetaEntry(dirent));
+    if (dirent == NULL) {
+      return false;
     }
 
     platform_stat64 info;
-    retval = platform_lstat(dirent->d_name, &info);
-    if (retval != 0) {
-      closedir(dirp);
+    if (platform_lstat(dirent->d_name, &info) != 0) {
       return false;
     }
 
-    if (S_ISDIR(info.st_mode)) {
-      // Recursion
-      retval = chdir(dirent->d_name);
-      if (retval != 0) {
-        closedir(dirp);
-        return false;
-      }
-      retval = ClearWorkingDir();
-      if (!retval) {
-        closedir(dirp);
-        return false;
-      }
-      retval = chdir("..");
-      if (retval != 0) {
-        closedir(dirp);
-        return false;
-      }
-      retval = rmdir(dirent->d_name);
-      if (retval != 0) {
-        closedir(dirp);
-        return false;
-      }
+    entry->is_directory = S_ISDIR(info.st_mode);
+    entry->name         = dirent->d_name;
+    return true;
+  }
+
+ protected:
+  string GetCurrentWorkingDirectory() {
+    char path[PATH_MAX];
+    const char* cwd = getcwd(path, PATH_MAX);
+    assert(cwd == path);
+    return string(cwd);
+  }
+
+  void ChangeDirectory(const string &path) {
+    const int retval = chdir(path.c_str());
+    assert(retval == 0);
+  }
+
+  bool IsMetaEntry(const platform_dirent64 *dirent) {
+    return (strcmp(dirent->d_name, ".")  == 0) ||
+           (strcmp(dirent->d_name, "..") == 0);
+  }
+
+ private:
+  const string  previous_path_;
+        DIR    *directory_handle_;
+};
+
+static bool ClearDirectory(const string &path) {
+  ScopedWorkingDirectory swd(path);
+  if (!swd) {
+    return false;
+  }
+
+  bool success = true;
+  ScopedWorkingDirectory::DirectoryEntry dirent;
+  while (success && swd.NextDirectoryEntry(&dirent)) {
```

Comment:
This stops the recursive delete as soon as there was an error, which might result in an unfinished cleanup. Whatever is left-over should be taken care of by the following publish. Not sure if we should have more safe guards here?